Advanced settings panel

jsarapata opened the pull request:
This change allows setting the console filtering level in both Firefox and Chrome to address Issue #1369. More specifically, the changes include:
1) Adding a consoleFilter field to our global settings, propagating it, and making it active.
2) Creating an advanced settings panel that gives full ability to edit the JSON of our global settings.
3) Removing the STUN server settings and advanced setting pane in our basic settings. These are subsumed by the advanced settings pane, which will come up when you click on advanced settings.
4) As long as I was editing settings.html, I replaced " by ' everywhere.
5) As long as I was editing uproxy_core_api.ts, I moved some type definitions to uProxy coding standards by removing the space in ': <type>'

There are a couple issues with the code.
1) The advanced settings panel comes up fine, but if you edit the text block at all, an extraneous scrollbar appears on the entire panel. If anybody knows how to fix this, I'd be grateful.
2) I'm not sure if this gets saved or if we we need a Restore Factory Settings capability.

The basic screen is:
![image](https://cloud.githubusercontent.com/assets/10854798/7596921/6d168cb6-f8be-11e4-8e24-5062d2ec8943.png)

After you set the text, it looks like(notice the extra scrollbar): 
![image](https://cloud.githubusercontent.com/assets/10854798/7596942/910fb1d8-f8be-11e4-8977-693c34b85115.png)

If you try entering bad JSON:
![image](https://cloud.githubusercontent.com/assets/10854798/7596969/c47248b0-f8be-11e4-9ddb-b24577597324.png)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1527)
<!-- Reviewable:end -->

jpevarnek commented:
I skipped over a lot of the CSS rules, so someone else should definitely take a look at the styling.

As a general thing, it seems worth considering what happens if `model.globalSettings` changes due to something else while you are editing (right now, nothing happens, and you may overwrite the changes, we should probably at least have a comment about that)

---

Reviewed files:
- src/generic_core/globals.ts @ r1
- src/generic_core/uproxy_core.ts @ r1
- src/generic_ui/polymer/advanced-settings.ts @ r1
- src/generic_ui/polymer/root.html @ r1
- src/generic_ui/polymer/settings.html @ r1
- src/generic_ui/polymer/settings.ts @ r1
- src/generic_ui/scripts/ui.ts @ r1
- src/interfaces/uproxy_core_api.ts @ r1

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8h9fmb9CmMzur9S_I)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L10)):</sup>
You do not need the empty attributes tag here

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 117 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kDY812bSTRHqt2KZ)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L117)):</sup>
This is more of an Izzie thing, but having a bit more differentiation between this and the success text would probably be useful (e.g. making it red)

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 8 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jcBXv1fskkgrYhLg)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L8)):</sup>
I don't think you need to look for either argument here (if a details object is passed in, it also would definitely not match that type)

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jthsx8iPJebW4q_W)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L10)):</sup>
Instead of manually toggling the hidden variables, it's a bit better (read: less prone to breaking if someone else pokes at the code) if you set another property of the polymer object to be true/false/an enum value and then determine whether the element is hidden based off of that. (e.g. `hidden ?= '{{ status !== StatusICareAbout }}'`)

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kU6dbM4WLnSrEpzi)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L16)):</sup>
Instead of replacing it completely, it might be better to use something like `_.merge`.  It might also be better to set `this.settings` to be the re-stringified settings afterwards.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 18 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jlkNfq4hiY-NGbC7)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L18)):</sup>
We generally put the catch on the same line as the closing brace (not sure if that is actually in the style guide though)

---

<sup>**[src/generic_ui/scripts/ui.ts, line 53 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8luoatKXTy7Sp5dJ5)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/scripts/ui.ts#L53)):</sup>
Can you pull in the actual enum value? (that or use 0)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

dborkan commented:
<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8mIWta_tXWHO_Bf7g)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L16)):</sup>
Can we also add a check that the user hasn't added or removed any keys from global settings?  Especially removing keys I think will cause bugs (e.g. values we expect to exist will be undefined)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

jsarapata commented:
I think this is pretty safe because we reset settings on overy open, but I added a comment at the top of the ts file.

---





<sup>**[src/generic_ui/polymer/advanced-settings.html, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8h9fmb9CmMzur9S_I)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L10)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 117 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kDY812bSTRHqt2KZ)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L117)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 8 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jcBXv1fskkgrYhLg)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L8)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jthsx8iPJebW4q_W)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L10)):</sup>
Done. I needed to created a toggleStatus because the hidden value didn't seem to update automagically when I changed the state.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kU6dbM4WLnSrEpzi)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L16)):</sup>
Done. Partway. I reset to the Jsonified string. I don't like the idea of merge because that wouldn't that create a monster object? It might make sense for adding STUN servers, but does the key check make it not necessary?

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8mIWta_tXWHO_Bf7g)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L16)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 18 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jlkNfq4hiY-NGbC7)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L18)):</sup>
Done.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 53 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8luoatKXTy7Sp5dJ5)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/scripts/ui.ts#L53)):</sup>
I tried using the enum value first, but I can't import it from the core. The 2 is there because that is the one that the core is actually initialized to. In any case, Daniel assures me that these values will be overwritten by a message from the core as part of start up.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed files:
- src/generic_ui/polymer/advanced-settings.ts @ r2

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 121 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-JpI9bMS8OSBkrAFmi4G)** ([raw file](https://github.com/uproxy/uproxy/blob/ce3bcdb6caa68c6026e7958fd01b162cd959859b/src/generic_ui/polymer/advanced-settings.html#L121)):</sup>
I'm actually a bit surprised this works, polymer calls for it to be set as `hidden?='{{ something }}'`

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 122 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kDY812bSTRHqt2KZ)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L122)):</sup>
I think this might be better having a `.error` or `.advanctedSettingsText.error` CSS rule and only specifying the color in that rule (and then applying both classes to the actual element)

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kU6dbM4WLnSrEpzi)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L16)):</sup>
The key check does make in unnecessary, however, I personally like it a lot better as a general idea for the UX.  If I wanted to tell someone just to change the log level, giving them the object `{ "consoleLevel": 0 }` and telling them to paste it in feels a lot easier than "Open up the advanced settings, find the log level property, change the number, and don't touch anything else".

Your call on this one, both things are looking at the same problem.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 37 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jthsx8iPJebW4q_W)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L37)):</sup>
You should not need that (actually, this is probably from the note I made above regarding `=` instead of `?=`)

---

<sup>**[src/generic_ui/scripts/ui.ts, line 53 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8luoatKXTy7Sp5dJ5)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/scripts/ui.ts#L53)):</sup>
`import loggingTypes = require('../../../../third_party/uproxy-lib/loggingprovider/loggingprovider.types');` did not work?  I cannot think of any reason that would not work and if it is failing, I think it's something we'd want to look into a bit more.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

jsarapata commented:
<sup>**[src/generic_ui/polymer/advanced-settings.html, line 121 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-JpI9bMS8OSBkrAFmi4G)** ([raw file](https://github.com/uproxy/uproxy/blob/ce3bcdb6caa68c6026e7958fd01b162cd959859b/src/generic_ui/polymer/advanced-settings.html#L121)):</sup>
I tried hidden?=... It still doesn't work. Looking at the docs: https://www.polymer-project.org/0.5/docs/polymer/binding-types.html, this doesn't set it to true or false, but controls whether the attribute appears at all. In any case, it doesn't actually toggle.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 122 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kDY812bSTRHqt2KZ)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L122)):</sup>
I'm not sure exactly what you were requesting, but I modified so all messages use advancedSettingsText, and specifically overrode the color on the two error ites,. So there is no more redundant size and padding information.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 37 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jthsx8iPJebW4q_W)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L37)):</sup>
I do need it, even with ?=

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 60 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kU6dbM4WLnSrEpzi)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L60)):</sup>
I see the use case, but I think it's more likely people will just modify the string that comes up. Given that I have been able to find no documentation on JSON.merge, or even see a description of where in the heck the JSON object comes from in a way that works with noth Chrome and Firefox, I'd rather just leave it.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 53 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8luoatKXTy7Sp5dJ5)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/scripts/ui.ts#L53)):</sup>
It does not:
>> Error: Cannot find module '../../../../third_party/uproxy-lib/loggingprovider/loggingprovider.types' from '/usr/local/google/home/jsarapata/git/uProxy/build/dev/uproxy/chrome/extension/generic_ui/scripts'
Warning: Error running grunt-browserify. Use --force to continue.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed files:
- src/generic_ui/polymer/root.html @ r3

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 119 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-JpI9bMS8OSBkrAFmi4G)** ([raw file](https://github.com/uproxy/uproxy/blob/ce3bcdb6caa68c6026e7958fd01b162cd959859b/src/generic_ui/polymer/advanced-settings.html#L119)):</sup>
The issue you are running into here is that you are trying to do a comparisong with StatusState which is not a property of the polymer object.  To get around that, you can add `StatusState: StatusState,` as one of the initialization properties.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 120 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kDY812bSTRHqt2KZ)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L120)):</sup>
I'm being a bit nit-picky here, but I think it would work better with them both sharing a second css class (e.g. `error`) instead of needing to specify the ID of every single block of text that should be treated as an error

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 37 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jthsx8iPJebW4q_W)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L37)):</sup>
See above.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 60 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kU6dbM4WLnSrEpzi)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L60)):</sup>
Sorry, `_.merge`, not `JSON.merge`.  Anyways, what you have works.

---

<sup>**[src/generic_ui/scripts/ui.ts, line 53 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8luoatKXTy7Sp5dJ5)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/scripts/ui.ts#L53)):</sup>
Ugh, this is a build rule problem that we were conveniently ignoring before (the relative paths of everything currently change between us doing typescript compilation and browserification).  Yeah, you can just leave this :(.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

jsarapata commented:
<sup>**[src/generic_ui/polymer/advanced-settings.html, line 119 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-JpI9bMS8OSBkrAFmi4G)** ([raw file](https://github.com/uproxy/uproxy/blob/ce3bcdb6caa68c6026e7958fd01b162cd959859b/src/generic_ui/polymer/advanced-settings.html#L119)):</sup>
Done. OK, this works.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 120 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8kDY812bSTRHqt2KZ)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.html#L120)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/advanced-settings.ts, line 37 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1527#-Jp8jthsx8iPJebW4q_W)** ([raw file](https://github.com/uproxy/uproxy/blob/e204eae1bd11209eb299395dcba8541d45feb39a/src/generic_ui/polymer/advanced-settings.ts#L37)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Everything I've looked at looks fine (I still have not really poked around the html yet, I'll try to look at that soon or someone else can)

---

Reviewed files:
- src/generic_ui/polymer/advanced-settings.ts @ r4, r5

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->

jsarapata commented:
@lucyhe, I found the issue with the spurious scrollbar. You are right that it comes from mirror text. The issue also occurs with feedback. To reproduce, go to submit feed back, and in the feedback panel, keep hitting enter. I need to do it 17 times. Then the extra scrollbar will pop up.

The fix is to change 'overflow: auto' to 'overflow: hidden' in the formContainer block. I'm happy to make this change as soon as I land this pull request, but won't complain if you do it first.

jpevarnek commented:
@jsarapata I'm reasonably sure that will breaks things if you shrink the window down below the expected height (best guess: you will not be able to see the bottom of the text box)

Is there any reason why you have an absolute height specified for the autogrowing text box?

jsarapata commented:
@jpevarnek, mainly because I didn't consider that people would resize the uProxy window. I futzed around with flex and couldn't get anything that behaved reasonably. I think this is a combination of my lack of CSS experience, and paper-autogrow-textarea being flaky. In particular, it doesn't create the mirror area until you type in the box, and the mirror area seems to be pretty flaky.

Assuming you don't mind, I am reverting to 'overflow: auto' and will file a separate bug against both feedback and advanced settings. My reasons are that it is primarily cosmetic, effects both components, and will probably take an hour or two of play, along with filing possible bugs against Polymer.

jsarapata commented:
@lucyhe @dborkan Can one of you review the HTML file? It is the last blocker. Thanks!

dborkan commented:
Just looked at the HTML/CSS and it looks fine.  I'm not sure exactly where you are at now with the textarea issus, but in general I'd suggest that if you have trouble with the paper-autogrow-textarea you just use a standard HTML textarea tag with a little bit of styling.  In theory the paper polymer elements are supposed to be great but in practice they often have bugs and seem like overkill for the styling we want.

---

Review status: 6 of 9 files reviewed, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1527)
<!-- Sent from Reviewable.io -->
